daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -532,23 +531,30 @@ private[lf] object Speedy {
             steps += 1
             println(s"$steps: ${PrettyLightweight.ppMachine(this)}")
           }
-          if (returnValue != null) {
-            val value = returnValue
-            returnValue = null
-            popTempStackToBase()
-            popKont().execute(value)
-          } else {
-            val expr = ctrl
-            ctrl = null
-            expr.execute(this)
+          control match {
+            case Control.WeAreUnset() =>
+              sys.error("**attempt to run a machine with unset control")
+            case Control.WeAreComplete() =>
+              sys.error("**attempt to run a complete machine")
+            case Control.WeAreHungry(res) =>
+              sys.error(s"**attempt to run a hungry machine (feed me first): $res")
+            case Control.Expression(exp) =>
+              setControl(Control.WeAreUnset())
+              setControl(exp.execute(this))
```

Comment:
I'm probably missing something but shouldn't this be the same as

```suggestion
              setControl(exp.execute(this))
```

?

Reply:
Sadly no, because of the way we currently use `SpeedyHungy` and `SpeedyComplete` exceptions for _normal_ control flow within speedy. :fearful: 

But now that we have a proper `Control` type, I plan to avoid the use of exceptions very soon... :grinning:

Reply:
aah 🤦, `exp.execute(this)` runs in a context where `control=Control.WeAreUnset()`. I'll blame side effects :)

> I plan to avoid the use of exceptions very soon... 😀

:fire: